File: HW1/Homework.java
/*
 * Assignment: Homework #1
 * Name: Trey Manuszak
 * ASU email: tmanusza@asu.edu
 * Course: CSE240 TTh 1030-1145
 * File Description: Class with various methods which initializes arrays, prints arrays,
 * sorts arrays using selection sort, and calculates a factorial recursively.
 * */

public class Homework {

    /**
     * Description: Take an array of characters and change odd indices to 'a', else 'b'.
     * @param  chArray is an array of integers.
     * */
    public static void initializeArray(char[] chArray) {
        for (int i = 0; i < chArray.length; i++) {
            if (i % 2 == 1) { //if index odd, replace with 'a', else 'b'
                chArray[i] = 'a';
            } else {
                chArray[i] = 'b';
            }
        }
    }

    /**
     * Description: Take an array of characters and print to the console using a for loop.
     * @param chArray is an array of characters.
     * */
    public static void printArray(char[] chArray) {
        for (int i = 0; i < chArray.length; i++) {
            System.out.print(chArray[i]);
            System.out.print(' ');
        }
        System.out.println();
    }

    /**
     * Description: Take an array of characters and sort in descending order using the selection sort method.
     * @param unsortedChars is an array of characters.
     * */
    public static void selectionSort(char[] unsortedChars) {
        int len = unsortedChars.length;

        for (int i = 0; i < len - 1; i++) {
            int maxIndex = i;

            //Find the index of the largest element still unsorted
            for (int j = i + 1; j < len; j++) {
                if (unsortedChars[j] > unsortedChars[maxIndex]) {
                    maxIndex = j;
                }
            }

            //Swap the largest element with the first element
            char temp = unsortedChars[maxIndex];
            unsortedChars[maxIndex] = unsortedChars[i];
            unsortedChars[i] = temp;
        }
    }

    /**
     * Description: A recursive method to calculate the factorial of an integer.
     * @param number is the integer to find the factorial of.
     * @return Returns the factorial of number, otherwise assume number is 0, which returns 1.
     * */
    public static int factorial(int number) {
        if (number > 2) {
            //Multiply by the factorial of n-1
            number *= factorial(number - 1);
        } else if (number == 0) {
            return 1; //In case number == 0
        }

        return number;
    }

    /**
     * Description: Method to test the above methods
     * */
    public static void main(String[] args) {
        char [] a = {'2', '5', '7', '9', '2', '3', '1', '7', '9', '0'};
        char [] b = {'p', 'q', '8', '6', '9', '5', '3', '1', '6'};
        int [] c = {4, 2, 5, 3, 1};

        // testing initializeArray
        printArray(a); // print: 2 5 7 9 2 3 1 7 9 0
        initializeArray(a);
        printArray(a); // print: b a b a b a b a b a

        // testing selectionSort
        printArray(b); // print: p q 8 6 9 5 3 1 6
        selectionSort(b);
        printArray(b); // print: q p 9 8 6 6 5 3 1

        // testing factorial
        System.out.println(factorial (5)); //print: 120
        System.out.println(factorial (c[0])); //print: 24
        System.out.println(factorial (c[3])); //print: 6
    }
}

File: HW4/homework.cpp
/*
 * Assignment: Homework #4
 * Name: Trey Manuszak
 * ASU email: tmanusza@asu.edu
 * Course: CSE240 TTh 1030-1145
 * File Description: Source file with various methods for arrays.
 * */

#include <iostream>
using namespace std;
#include "homework.h"

/*
Constructor
--------------------
*/
Homework::Homework() {  

}

/*
Function:  initializingArray
--------------------
Function to initialize even indicies to 1 and odd indices to 0.

Parameters:
	int* :  Pointer to the array.
	int  :  Size of the array.		  
*/
void Homework::initializingArray(int* arr, int size) {
	for (int i = 0; i < size; i++) {
		if (i % 2 == 0) {
			arr[i] = 1;
		} else {
			arr[i] = 0;
		}
	}
}

/*
Function:  printingArray
--------------------
Function to print an array of ints.

Parameters:
	int* :  Pointer to the array.
	int  :  Size of the array. 
*/
void Homework::printingArray(int* arr, int size) {
	for (int i = 0; i < size - 1; i++) {
		cout << arr[i] << ", ";
	}
	cout << arr[size - 1] << endl;
}

/*
Function:  selectionSort
--------------------
Function to sort an array in ascending order using selection sort.
Parameters:
	int* :  Pointer to the array.
	int  :  Size of the array.
*/
void Homework::selectionSort(int* arr, int size) {
	int smallIndex = 0;
	// Shrinking window
	for (int i = 0; i < (size - 1); i++) {
		smallIndex = i;
		// Find smallest index
		for (int j = i + 1; j < size; j++) {
			if (arr[j] < arr[smallIndex]) {
				smallIndex = j; // New smallest index
			}
		}

		// Swap
		int temp = arr[smallIndex];
		arr[smallIndex] = arr[i];
		arr[i] = temp;
	} 	
}

/*
Function:  factorial
--------------------
Function to calculate the factorial of an integer recursively.
Parameters:
	int  :  The factorial to be calculated.		   
Returns:
    	int  :  The factorial of the integer.
*/
int Homework::factorial(int num) {
	if (num >= 2) {
		return num * factorial(num - 1);
	} else {
		return 1;
	}
}

